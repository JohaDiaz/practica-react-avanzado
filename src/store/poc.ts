
// function createStore(reducer){
//     let state;
//     let listeners = [];

//     function getState(){
//         return state;
//     }

//     function dispatch(action){
//         state = reducer(state, action);
//         listeners.forEach((l)=> l());
//     }

//     function subscribe(listener){
//         listeners.push(listener);
//         return function unsubscribe(){
//             listeners =listeners.filter((l) => l !== listener)
//         };
//     }
//     dispatch({type: "INIT"});

//     return{
//         getState,
//         dispatch,
//         subscribe,
//     };
// }

import { createStore } from "redux";

const initialState = 0; 

type Increment = {
    type: "increment";
    payload: number;
};

type Decrement = {
    type: "decrement";
};

type Actions = Increment | Decrement; 

function reducer(state = initialState, action: Actions){

    switch (action.type){
        case "increment":
            return state + action.payload;
        case "decrement":
            return state - 1; 
        default:
            return state; 
    }
}

export const store = createStore(reducer)

const render = () => console.log(store.getState());

const unsubscribe = store.subscribe(render);

render();

store.dispatch({type: "increment", payload:2});
store.dispatch({type: "increment", payload:3});
store.dispatch({type:"decrement"});
unsubscribe();
store.dispatch({type: "increment", payload:0});