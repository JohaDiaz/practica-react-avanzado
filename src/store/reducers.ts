//import { combineReducers } from "redux";
import type { Advert } from "../pages/adverts/types";
import type { Actions } from "./actions";

export type State = {
  auth: boolean;
  adverts: { data: Advert[] | null; loaded: boolean };
  ui: {
    pending: boolean;
    error: Error | null;
  };
  tags: string[];
  //adverts: { data:Advert[] | null; loaded: boolean};
};

const defaultState: State = {
  auth: false,
  adverts: { data: [], loaded: false },
  ui: {
    pending: false,
    error: null,
  },
  tags: [],
};

export function auth(
  state = defaultState.auth,
  action: Actions,
): State["auth"] {
  switch (action.type) {
    case "auth/login/fulfilled":
      return true;
    case "auth/logout":
      return false;
    default:
      return state;
  }
}

export function adverts(
  state = defaultState.adverts,
  action: Actions,
): State["adverts"] {
  switch (action.type) {
    case "adverts/loaded/fulfilled":
      console.log("Payload recibido:", action.payload);
      return {
        data: action.payload.data,
        loaded: action.payload.loaded,
      };
    case "adverts/created/fulfilled":
      return { ...state, data: [...(state.data ?? []), action.payload] };
    case "adverts/deleted":
      return {
        ...state,
        data:
          state.data?.filter((advert) => advert.id !== action.payload) ?? null,
      };
    default:
      return state;
  }
}

export function ui(state = defaultState.ui, action: Actions): State["ui"] {
  switch (action.type) {
    case "ui/reset-error":
      return { ...state, error: null };
    case "auth/login/pending":
      return { pending: true, error: null };
    case "auth/login/fulfilled":
      return { pending: false, error: null };
    case "auth/login/rejected":
      return { pending: false, error: action.payload };
    default:
      return state;
  }
}

export function tags(state = [], action: Actions): string[] {
  switch (action.type) {
    case "tags/loaded":
      return action.payload;
    default:
      return state;
  }
}

// export function reducer(state =defaultState, action: Actions): State{

//     return {
//         auth: auth(state.auth, action),
//         adverts: adverts(state.adverts, action),
//     }
// };

//export const reducer = combineReducers({ auth: auth, adverts: adverts});

//export const reducer = combineReducers({ auth, adverts});

// export function reducer(state = defaultState, action: Actions):  State {
//     switch (action.type){
//         case 'auth/login':
//             return {...state, auth: true};
//         case 'auth/logout':
//             return {...state, auth: false};
//         case 'adverts/created':
//             return { ...state, adverts: [...state.adverts, action.payload] };
//         case 'adverts/loaded':
//             return { ...state, adverts: action.payload };
//         default:
//             return state;
//     }
// }
