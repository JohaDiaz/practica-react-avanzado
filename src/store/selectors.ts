import type { RootState } from ".";

export const getIsLogged = (state: RootState) => state.auth;

export const getAllAdverts = (state: RootState) => state.adverts;

//export const getTweets = (state: RootState) => state.tweets.data || [];

export const getAdvertDetail = (state: RootState, advertId?: string) =>
    advertId
      ? state.adverts.find((advert) => String(advert.id).trim() === String(advertId).trim()) ?? null
      : null;