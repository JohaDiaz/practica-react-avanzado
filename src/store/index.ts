import { combineReducers, legacy_createStore as createStore } from "redux";
import { composeWithDevTools } from "@redux-devtools/extension";
import { useDispatch, useSelector } from "react-redux";
import { auth, adverts, ui } from "./reducers";
// Importa un objeto con todos los exports del fichero


import type { State } from "./reducers";


export default function configureStore(preloadedState?: Partial<State>){
    const rootReducer = combineReducers({ auth, adverts, ui });
    const store = createStore(
        rootReducer,
        //@ts-expect-error: preloaded state is not inferred
        preloadedState,
        composeWithDevTools(),
         );
         console.log("Estado inicial de Redux:", store.getState());

    return store;
    
}


export type AppStore = ReturnType<typeof configureStore>;
export type AppGetState = AppStore["getState"];
export type RootState = ReturnType<AppGetState>;
export type AppDispatch = AppStore["dispatch"];


export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();