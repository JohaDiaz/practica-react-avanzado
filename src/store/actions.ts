import type { Credentials } from "@/pages/auth/types";
import type { Advert, CreateAdvertDto } from "../pages/adverts/types";
import { isApiClientError } from "@/api/error";
import type { AppThunk } from ".";
import { getAdvertDetail } from "@/store/selectors";

//login
type AuthLoginPending = {
  type: "auth/login/pending";
};

type AuthLoginFulfilled = {
  type: "auth/login/fulfilled";
};

type AuthLoginRejected = {
  type: "auth/login/rejected";
  payload: Error;
};

type AuthLogout = {
  type: "auth/logout";
};

type AdvertsLoadedFulfilled = {
  type: "adverts/loaded/fulfilled";
  payload: { data: Advert[]; loaded: boolean };
};

type AdvertCreatedFulfilled = {
  type: "adverts/created/fulfilled";
  payload: Advert;
};

type AdvertsDetailRejected = {
  type: "adverts/detail/rejected";
  payload: Error;
};

type AdvertDeleted = {
  type: "adverts/deleted";
  payload: string;
};

type TagsLoaded = {
  type: "tags/loaded";
  payload: string[];
};

type UiResetError = {
  type: "ui/reset-error";
};

//actions creators
export const authLoginPending = (): AuthLoginPending => ({
  type: "auth/login/pending",
});

export const authLoginFulfilled = (): AuthLoginFulfilled => ({
  type: "auth/login/fulfilled",
});

export const authLoginRejected = (error: Error): AuthLoginRejected => ({
  type: "auth/login/rejected",
  payload: error,
});

//api
export function authLogin(
  credentials: Credentials,
  remember: boolean,
): AppThunk<Promise<void>> {
  return async function (dispatch, _getState, { api, router }) {
    dispatch(authLoginPending());
    try {
      await api.authService.login(credentials, remember);
      dispatch(authLoginFulfilled());
      const to = router.state.location.state?.from ?? "/";
      router.navigate(to, { replace: true });
    } catch (error) {
      if (isApiClientError(error)) {
        dispatch(authLoginRejected(error));
      }
    }
  };
}

export const authLogout = (): AuthLogout => ({
  type: "auth/logout",
});

export const AdvertsLoadedFulfilled = (
  adverts: Advert[],
  loaded?: boolean,
): AdvertsLoadedFulfilled => {
  return {
    type: "adverts/loaded/fulfilled",
    payload: { data: adverts, loaded: !!loaded },
  };
};

// export const AdvertsLoadedFulfilled = (
//   adverts: Advert[],
// ): AdvertsLoadedFulfilled => {
//   return {
//     type: "adverts/loaded/fulfilled",
//     payload: adverts,
//   };
// };

//api
export const advertsLoaded = (): AppThunk<Promise<void>> => {
  return async (dispatch, getState, { api }) => {
    const state = getState();
    if (state.adverts.loaded) {
      return;
    }

    try {
      const adverts = await api.advertsService.getAdverts();
      dispatch(AdvertsLoadedFulfilled(adverts, true));
    } catch (error) {
      console.error(error);
    }
  };
};

export const advertsDetailRejected = (error: Error): AdvertsDetailRejected => ({
  type: "adverts/detail/rejected",
  payload: error,
});

//api
export const advertLoaded = (advertId: string): AppThunk<Promise<void>> => {
  return async (dispatch, getState, { api }) => {
    const state = getState();
    if (getAdvertDetail(state, advertId)) {
      return;
    }

    try {
      const advert = await api.advertsService.getAdvert(advertId);
      dispatch(AdvertsLoadedFulfilled([advert]));
    } catch (error) {
      if (isApiClientError(error)) {
        dispatch(advertsDetailRejected(error));
      }
      console.error(error);
    }
  };
};

export const advertCreatedFulfilled = (
  advert: Advert,
): AdvertCreatedFulfilled => ({
  type: "adverts/created/fulfilled",
  payload: advert,
});

//api
export function advertCreate(
  createAdvertDto: CreateAdvertDto,
): AppThunk<Promise<Advert>> {
  return async function (dispatch, _getState, { api, router }) {
    // AdverCreatePending
    try {
      const createdAdvert =
        await api.advertsService.createAdvert(createAdvertDto);
      const advert = await api.advertsService.getAdvert(
        createdAdvert.id.toString(),
      );
      dispatch(advertCreatedFulfilled(advert));
      await router.navigate(`/adverts/${createdAdvert.id}`);
      return advert;
    } catch (error) {
      if (isApiClientError(error)) {
        // Manage advertCreateRejected
      }
      throw error;
    }
  };
}

export const advertDeleted = (advertId: string): AdvertDeleted => ({
  type: "adverts/deleted",
  payload: advertId,
});

export const tagsLoaded = (tags: string[]): TagsLoaded => ({
  type: "tags/loaded",
  payload: tags,
});

export const uiResetError = (): UiResetError => ({
  type: "ui/reset-error",
});

// thunk para cargar los tags desde la API
// api
export const loadTags = (): AppThunk<Promise<void>> => {
  return async (dispatch, getState, { api }) => {
    const state = getState();
    if (state.tags.length) return;

    try {
      const tags = await api.advertsService.getTags();
      dispatch(tagsLoaded(tags));
    } catch (error) {
      console.error("Error al cargar tags:", error);
    }
  };
};

export type Actions =
  | AuthLoginPending
  | AuthLoginFulfilled
  | AuthLoginRejected
  | AuthLogout
  | AdvertsLoadedFulfilled
  | AdvertCreatedFulfilled
  | AdvertsDetailRejected
  | AdvertDeleted
  | TagsLoaded
  | UiResetError;
