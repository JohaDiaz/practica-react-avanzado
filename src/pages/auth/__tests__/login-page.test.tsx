import { render, screen, fireEvent } from "@testing-library/react";
import LoginPage from "../login-page";
import { MemoryRouter } from "react-router-dom";
import { vi } from "vitest";
import * as store from "@/store";
import * as actions from "@/store/actions";

// Mock useAppDispatch y useAppSelector
const mockDispatch = vi.fn();
vi.mock("@/store", async () => {
  const actual = await vi.importActual<typeof store>("@/store");
  return {
    ...actual,
    useAppDispatch: () => mockDispatch,
    useAppSelector: vi.fn(() => ({ pending: false, error: null })),
  };
});

// Mock de la acciÃ³n authLogin
vi.mock("@/store/actions", async () => {
  const actual = await vi.importActual<typeof actions>("@/store/actions");
  return {
    ...actual,
    authLogin: vi.fn(() => ({ type: "MOCKED_AUTH_LOGIN" })),
  };
});

describe("LoginPage component", () => {
  test("calls authLogin when form is submitted", async () => {
    render(
      <MemoryRouter>
        <LoginPage />
      </MemoryRouter>,
    );

    const emailInput = screen.getByPlaceholderText("Email");
    const passwordInput = screen.getByPlaceholderText("Password");
    const submitButton = screen.getByRole("button", {
      name: /Log in to Nodepop/i,
    });

    fireEvent.change(emailInput, { target: { value: "test@email.com" } });
    fireEvent.change(passwordInput, { target: { value: "123456" } });
    fireEvent.click(submitButton);

    // Espera un tick del event loop
    await new Promise((r) => setTimeout(r, 0));

    expect(mockDispatch).toHaveBeenCalled();
    expect(actions.authLogin).toHaveBeenCalledWith(
      { email: "test@email.com", password: "123456" },
      false,
    );
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <LoginPage />
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
