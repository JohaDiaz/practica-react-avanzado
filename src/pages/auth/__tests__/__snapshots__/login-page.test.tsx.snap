// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="mx-auto h-dvh max-w-md"
  >
    <div
      class="grid gap-8 px-6 py-6 pt-12"
    >
      <header
        class="grid justify-items-center gap-4"
      >
        <svg
          class="h-24 w-24"
          color="var(--primary)"
          viewBox="0 0 200 200"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="none"
            height="140"
            rx="20"
            stroke="currentColor"
            stroke-width="8"
            width="180"
            x="10"
            y="40"
          />
          <path
            d="M55 40 Q100 0 145 40"
            fill="none"
            stroke="currentColor"
            stroke-width="8"
          />
          <g
            stroke="currentColor"
            stroke-width="5"
            transform="translate(100, 110) scale(1.6)"
          >
            <circle
              cx="0"
              cy="0"
              fill="currentColor"
              r="5"
            />
            <ellipse
              fill="none"
              rx="30"
              ry="12"
              transform="rotate(0)"
            />
            <ellipse
              fill="none"
              rx="30"
              ry="12"
              transform="rotate(60)"
            />
            <ellipse
              fill="none"
              rx="30"
              ry="12"
              transform="rotate(120)"
            />
          </g>
        </svg>
        <h1
          class="text-center text-3xl font-bold sm:text-4xl"
        >
          Log in to Nodepop
        </h1>
      </header>
      <div>
        <form
          class="grid gap-4"
        >
          <label
            class="text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 grid w-full items-center gap-2"
            data-slot="label"
          >
            Email
            <input
              autocomplete="off"
              class="border-input file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 aria-invalid:outline-destructive/60 dark:aria-invalid:outline-destructive dark:aria-invalid:ring-destructive/40 aria-invalid:ring-destructive/20 aria-invalid:border-destructive/60 dark:aria-invalid:border-destructive flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium focus-visible:ring-4 focus-visible:outline-1 disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 aria-invalid:focus-visible:ring-[3px] aria-invalid:focus-visible:outline-none md:text-sm dark:aria-invalid:focus-visible:ring-4"
              data-slot="input"
              name="email"
              placeholder="Email"
              type="email"
              value=""
            />
          </label>
          <label
            class="text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 grid w-full items-center gap-2"
            data-slot="label"
          >
            Password
            <input
              autocomplete="off"
              class="border-input file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 aria-invalid:outline-destructive/60 dark:aria-invalid:outline-destructive dark:aria-invalid:ring-destructive/40 aria-invalid:ring-destructive/20 aria-invalid:border-destructive/60 dark:aria-invalid:border-destructive flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium focus-visible:ring-4 focus-visible:outline-1 disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 aria-invalid:focus-visible:ring-[3px] aria-invalid:focus-visible:outline-none md:text-sm dark:aria-invalid:focus-visible:ring-4"
              data-slot="input"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </label>
          <label
            class="text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 w-full items-center gap-2 flex py-2"
            data-slot="label"
          >
            <button
              aria-checked="false"
              class="peer data-[state=checked]:bg-primary data-[state=unchecked]:bg-input ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 inline-flex h-5 w-9 shrink-0 items-center rounded-full border-2 border-transparent shadow-xs transition-[color,box-shadow] focus-visible:ring-4 focus-visible:outline-hidden focus-visible:outline-1 disabled:cursor-not-allowed disabled:opacity-50 aria-invalid:focus-visible:ring-0"
              data-slot="switch"
              data-state="unchecked"
              role="switch"
              type="button"
              value="remember"
            >
              <span
                class="bg-background pointer-events-none block size-4 rounded-full ring-0 shadow-lg transition-transform data-[state=checked]:translate-x-4 data-[state=unchecked]:translate-x-0"
                data-slot="switch-thumb"
                data-state="unchecked"
              />
            </button>
            <input
              aria-hidden="true"
              name="remember"
              style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
              tabindex="-1"
              type="checkbox"
              value="remember"
            />
            Remember me next time
          </label>
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 [&_svg]:shrink-0 ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 focus-visible:ring-4 focus-visible:outline-1 aria-invalid:focus-visible:ring-0 bg-primary text-primary-foreground shadow-sm hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 w-full"
            data-slot="button"
            disabled=""
          >
            Log in to Nodepop
          </button>
        </form>
        <section
          aria-atomic="false"
          aria-label="Notifications alt+T"
          aria-live="polite"
          aria-relevant="additions text"
          tabindex="-1"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
